nsmdp/nsmdp.go

```diff
@@ -20,89 +20,123 @@
-	*pluginapi.Device
-	token string
+type nsmClientEndpoints struct {
+	nsmSockets map[string]nsmSocket
+	logger     logging.PluginLogger
 }
 
-const (
-	resourceName    = "nsm.ligato.io/socket"
-	serverSock      = pluginapi.DevicePluginPath + "nsm.ligato.io.sock"
-	initDeviceCount = 10
-)
+// Define functions needed to meet the Kubernetes DevicePlugin API
+func (n *nsmClientEndpoints) GetDevicePluginOptions(context.Context, *pluginapi.Empty) (*pluginapi.DevicePluginOptions, error) {
+	n.logger.Infof("GetDevicePluginOptions was called.")
+	return &pluginapi.DevicePluginOptions{}, nil
+}
 
-func NewNSMDevicePlugin() *NSMDevicePlugin {
-	n := &NSMDevicePlugin{
-		DevicePlugin: deviceplugin.NewDevicePlugin(serverSock, resourceName),
-		devs:         make(map[string]*NSMDevice),
-		stop:         make(chan interface{}),
-		updatedevs:   make(chan *NSMDevice, 10),
-	}
-	for i := uint(0); i < initDeviceCount; i++ {
-		dev := &NSMDevice{
-			Device: &pluginapi.Device{
-				ID:     "NSM_" + string(i),
-				Health: pluginapi.Healthy,
-			},
-			token: generateToken(),
-		}
-		n.updatedevs <- dev
+func startClientServer(client *nsmSocket, logger logging.PluginLogger) {
+	listenEndpoint := client.socketPath
+	fi, err := os.Stat(listenEndpoint)
+	if err == nil && (fi.Mode()&os.ModeSocket) != 0 {
+		os.Remove(listenEndpoint)
+	}
+	if err != nil && !os.IsNotExist(err) {
+		logger.Errorf("failure stat of socket file %s with error: %+v", client.socketPath, err)
+		client.allocated = false
+		return
 	}
-	return n
-}
 
-func generateToken() string {
-	b := make([]byte, 16)
-	rand.Read(b)
-	return fmt.Sprintf("%x", b)
-}
+	unix.Umask(socketMask)
+	sock, err := net.Listen("unix", listenEndpoint)
+	if err != nil {
+		logger.Errorf("failure to listen on socket %s with error: %+v", client.socketPath, err)
+		client.allocated = false
+		return
+	}
+	grpcServer := grpc.NewServer([]grpc.ServerOption{}...)
+
+	logger.Infof("Starting Client gRPC server listening on socket: %s", serverSock)
+	go grpcServer.Serve(sock)
 
-func (n *NSMDevicePlugin) Stop() error {
-	err := n.DevicePlugin.Stop()
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	conn, err := dial(ctx, listenEndpoint)
 	if err != nil {
-		return err
+		logger.Errorf("failure to communicate with the socket %s with error: %+v", client.socketPath, err)
+		client.allocated = false
 	}
-	close(n.stop)
-	return nil
+	defer conn.Close()
+	defer cancel()
+	logger.Infof("Client Server socket: %s is operational", listenEndpoint)
 
+	// Wait for shutdown
+	select {
+	case <-client.stopChannel:
+		logger.Infof("Server for socket %s received shutdown request", client.socketPath)
+	}
+	client.allocated = false
+	client.stopChannel <- true
 }
 
-func (n *NSMDevicePlugin) Allocate(ctx context.Context, reqs *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
+func (n *nsmClientEndpoints) Allocate(ctx context.Context, reqs *pluginapi.AllocateRequest) (*pluginapi.AllocateResponse, error) {
```

Comment:
Do we know if Allocate can be called concurrently with another Allocate or if ListAndWatch can read or update state concurrently while Allocate is running?

We should consider synchronizing this code and the hot spot in ListAndWatch.

Reply:
Allocate call is not concurrent. ListAndWatch for out case is not really required, I mean running in the continues loop, as we do not expect to see any changes in the number or state of the pool os socket. In theory I could reply just once, and then do a fake loop just to keep grpc stream connection alive.